docs/proposals/device-crd.md

```diff
@@ -0,0 +1,207 @@
+
+### Modelling Devices as CRD’s
+<img src="../images/device-crd/device-crd-model.png">
+
+[Custom resources](https://kubernetes.io/docs/concepts/extend-kubernetes/api-extension/custom-resources/) allows us to define declarative APIs for custom resources. Devices attached to edge nodes can be described using a custom resource and managed using APIs installed in a Kubernetes cluster. Users can interact with the device CRD’s using standard `kubectl` operations or using client libraries. The Device controller is a CRD controller whose job is to ensure the device updates are synced between the cloud and the edge node.
+
+### Device Custom Resource Overview
+The device custom resource layout is shown below. It contains standard fields ( Name, Namespace, Labels , Annotations etc. ) which help the Kubernetes ecosystem to manage resources. In addition, it contains custom fields in the Spec and Status detailed further below.
+<img src="../images/device-crd/device-crd.png">
+
+The `Spec` contains the following fields
+
+| Field                 | Description                                                                      |
+|-----------------------|----------------------------------------------------------------------------------|
+| Attributes            | A list of device model attributes. Each device attribute describes the data type ,<br />permissible values, defaults etc.|
+| Attribute Visitors    | The attribute visitors describe how to access the attribute values. <br />Each attribute has a visitor. |
+| Connection Type       | The connection type field indicates whether the device is directly connected <br /> to the cloud ( active devices ),or the device is connected via the edge platform(i.e indirect connection for passive devices). |
+| Node                  | The edge node details to which this device belongs to.|
+| Protocol              | The protocol supported by the device to communicate (e.g BLE, Modbus, OPC UA). <br />The configuration can contain more details about the connection mode , server, port details etc.|
+
+The `Status` contains the following fields
+
+| Field                 | Description                                                                      |
+|-----------------------|----------------------------------------------------------------------------------|
+|Twin                   | A list of device twin attributes.Each device twin attribute is like a control attribute which can be managed from <br />the cloud or the edge. Currently, the cloud configures the `Expected`state of a device and this configuration is pushed to the edge node. The mapper sends a command to the device to change as per the expected state . It receives the Actual state of the attribute once the previous operation is complete and sends the actual state to the cloud. Offline device interaction in the edge is possible via twin attributes for control/command operations.|
```

Comment:
There are some device properties which can be controlled ( accessMode : rw ). It would make sense to have expected and actual device twin attributes for such properties . If a device property is read-only like temperture from a sensor ( accessMode: r) , then it won't be logical to have an "expected" state for it. 

The device twin attributes in the DeviceStatus just report the current expected and actual values for a control device property. The property name in the spec should match the device twin name in the status.